src/app/vendors/actions.ts

```diff
@@ -23,22 +39,149 @@ export function loadAllVendors(
+          const vendorHash = Number(vendorHashKey);
+          // if it's not one of the vendors the api says to include on the vendors page,
+          // we don't need its stats. right now this mainly serves to exclude yuna
+          if (!displayVendors.includes(vendorHash)) {
+            return;
+          }
+          // if we find an item that needs components, include this vendor
+          for (const { itemHash } of Object.values(sales.saleItems)) {
+            if (itemNeedsComponents(defs, itemHash)) {
+              return vendorHash;
+            }
+          }
+        },
+      );
+      timings.componentsNeeded = vendorsNeedingComponents.length;
+
+      const { collapsedSections } = settingsSelector(getState());
+
+      // decide the order of the single-vendor requests (false is earlier in sort than true)
+      vendorsNeedingComponents.sort(
+        chainComparator(
+          // prioritize vendors in groups (meaning top-level, not click-in sub-vendors)
+          compareBy((h) => !defs.Vendor.getOptional(h)?.groups.length),
+          // deprioritize vendors whose sections are collapsed on the vendors page
+          compareBy((h) => Boolean(collapsedSections[`d2vendor-${h}`])),
+          // sort by their position on the page lol
+          compareBy((h) => displayVendors.indexOf(h)),
+        ),
+      );
+
+      // TO-DO ↑: limit collapsedSections prioritization to when we're on vendors page?
+      // TO-DO ↑: if the current page is optimizer/loadouts:
+      //          prioritize armor? do all single-vendor fetches then submit their
+      //          itemComponents en-masse, to cause only one optimizer recalculation?
+
+      for (const vendorHash of vendorsNeedingComponents) {
+        start = Date.now();
+        const vendorResponse = await getVendorSaleComponents(account, characterId, vendorHash);
+        timings.componentsTime += Date.now() - start;
+        timings.componentsFetched++;
+        dispatch(loadedSingle({ vendorResponse, characterId, vendorHash }));
+      }
     } catch (e) {
       const error = convertToError(e);
       dispatch(loadedError({ characterId, error }));
+    } finally {
+      infoLog(
+        'Vendors',
+        [
+          `sales data loaded: ${timings.salesTime / 1000}s`,
+          `component data for ${timings.componentsFetched} vendors loaded: ${timings.componentsTime / 1000}s`,
+          timings.componentsFetched === timings.componentsNeeded
+            ? 0
+            : `${timings.componentsNeeded - timings.componentsFetched} components not loaded?`,
+        ]
+          .filter(Boolean)
+          .join(),
+      );
     }
   };
 }
+
+function itemNeedsComponents(defs: D2ManifestDefinitions, itemHash: number) {
+  const item = defs.InventoryItem.getOptional(itemHash);
+  return (
+    item &&
+    // probably just need this for weapon perks and armor stats?
+    // TO-DO: what about bounties and quests? where's their pre-existing progress stored? do we care?
```

Comment:
It might affect whether bounties appear purchase-able?

Reply:
bounties appear correct in my testing